flask_praetorian/base.py

```diff
@@ -721,25 +740,46 @@ def read_token_from_cookie(self):
         return self._unpack_cookie(flask.request.cookies)
 
     def read_token(self):
-        exc = None
-        if 'header' in self.jwt_places:
-            try:
-                return self.read_token_from_header()
-            except MissingToken as e:
-                exc = e
-        if 'cookie' in self.jwt_places:
+        """
+        Tries to unpack the token from the current flask request
+        in the locations configured by JWT_PLACES.
+        Check-Order is defined by the value order in JWT_PLACES.
+        """
+
+        for place in self.jwt_places:
             try:
-                return self.read_token_from_cookie()
-            except MissingToken as e:
-                exc = e
-        if exc:
-            raise MissingToken("JWT token not found in {}".format(
-                self.jwt_places))
+                return getattr(
+                    self,
+                    "read_token_from_{place}".format(
+                        place=place.lower(),
+                    ),
+                )()
+            except MissingToken:
+                pass
+            except AttributeError:
+                flask.current_app.logger.warning(
+                    "Flask_Praetorian hasn't implemented reading JWT tokens from location {place}. Please reconfigure JWT_PLACES.".format(  # noqa: E501
+                        place=place.lower(),
+                    )
+                )
+                flask.current_app.logger.warning(
+                    "Values accepted in JWT_PLACES are: {def_places}".format(
+                        def_places=DEFAULT_JWT_PLACES,
+                    )
+                )
+
+        raise MissingToken(
+            "Could not find token in any of the given locations: {locations}".format(  # noqa: E501
```

Comment:
use dedent if the line is too long.

Reply:
used dedent for this, but had to use .replace("\n", "") at the end to avoid \n inconsistencies

Reply:
the inconsistencies make the tests fail. uh oh

Reply:
a tab sneaked in somehow after I ran the tests on my machine... the tests did run through now